libs/depgraph/README.md

```diff
@@ -0,0 +1,296 @@
+Dependency Graph (depgraph)
+===========================
+
+Motivation
+----------
+
+Configuration management agents (like EVE's microservices) typically receive a declarative
+description of the intended (also called desired) state to apply into the managed system.
+This description is usually at a higher level of abstraction and expresses the intent,
+not the implementation. The agent then *renders* a corresponding set of low-level run-time
+objects (e.g. processes to run, files to write, config primitives to create), that, once
+*applied* into the running system, will implement the requested intent. The process of
+updating the system state is not trivial, however. The agent first have to compare the new
+intended state with the actual (also known as the current) state of the system. Based on
+the difference, the agent determines the sequence of operations to execute in order to
+synchronize the system state with the new intent. The order of operations often matters
+as it has to respect all dependencies that exist between configured objects. Operations
+are usually described using the CRUD interface (Create, Update/Modify, Read and Delete),
+but a different semantic for state manipulation is also possible.
+
+To summarize, the problem described above can be split into the following tasks:
+
+  1. From high-level declarative description (received from a user/controller) render
+     the corresponding set of configuration primitives, processes or files to create/run/write
+     in the running system.
+  2. Determine the difference between the current system state and the new (rendered)
+     intended state.
+  3. Determine the set of (Create/Modify/Delete) operations to execute and find a valid ordering.
+  4. Execute operations, collect any errors and report back to the user/controller.
+
+While the task no. 1 is specific to each management agent, tasks 2. and 3. (and to some degree
+also 4.) can be tackled in a generic way and in one place using a [dependency graph](https://en.wikipedia.org/wiki/Dependency_graph),
+implemented by this package.
+
+Features
+--------
+
+Depgraph implements a dependency graph solving the problem of the current ↔ intended state
+reconciliation as described above. Every configuration primitive (OS config, file, process, etc.),
+further denoted as *configuration item* or just *item*, can be represented by a single graph node.
+In the graph's API, this is described by the Item interface (see [depgraph_api.go](./depgraph_api.go))
+Dependencies between items are modeled using directed edges. For the start, the graph supports
+dependency with the semantics *"must exist"*. To give an example from networking: if a route depends
```

Comment:
'must exist' might not be sufficient, because the vlan link might exist before a change and exist after the change, but if it was deleted and recreated then the routes referencing it will have to be recreated. 'must exist and unchanged' would capture that.

I don't know if it is worth-while to try to distiguish between modifications to an parent which does vs. does not require the children to redo/recreate things, or just assume that if anything changes in a parent then the children will have to at least check in their modify/change action whether they need to do something.

Reply:
Note that "must exist" is a dependency applied to the entire lifetime of an item, not just prior to Create. So if a route depends on a VLAN link and the VLAN link is being re-created, it means that route will get deleted before the VLAN is, and then the route is created back after the VLAN is recreated. In other words, while the VLAN link is deleted (even if it is just "momentarily"), the route must be deleted as well.
"must exist" here means the route cannot *exist* without the VLAN.

Also, even if modification is done in-place (not by recreate), it is possible to [explicitly request re-create of items that depend on it](https://github.com/milan-zededa/eve/blob/dependency-graph/libs/depgraph/depgraph_api.go#L205-L207). This may be needed in some cases.